### ISSUE TITLE
What is the recommended way to handle http request error?

### ISSUE THREAD

**Issue by DevanXU** [redux-model/redux-model#6]
Version: `redux-model/taro@6.7.9`

We're using HttpService inside redux-model/taro to send request.
Our question is, what is the best way to handle request errors such as 502/500/...

In our case, when encountering 502, we received a response with:
- httpStatus: `undefined`

- response: unparsed html text like 
```
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.15.8</center>
</body>
</html>
```

- other info

The only way to get 502 info is to parse html text and extract it. Are we supposed to parse the response text and extract httpStatus info?

Thanks.
Devan

**Comment by fwh1990** [comment]
I think the property `httpStatus` will always have value when the request has response behavior. Can you provide taro info? And This issue will be solved faster if you can create simple repo for me.

**Comment by fwh1990** [comment]
It's unnecessary to parse html text here, exactly, we should get httpStatus from response headers.

**Comment by DevanXU** [comment]
Hi:

Sorry for late. Got stuck in the project ...

It's kinda difficult to extract demo code from my project, but I tried to fix the code in my local npm package. Refer as below code snippet comments (2 lines marked).
After the 2 line changes, we can get correct httpstatus and other error info now. From the debugging, `error.statusCode` is undefined.
Hope this is helpful for your official fix. Looking forward to it. :)

```var promise = task
      .then(function (response) {
        if (response.statusCode < 200 || response.statusCode >= 300 || (_this.config.isSuccess && !_this.config.isSuccess(response))) {
          return Promise.reject(response);
        }
        if (_this.config.transformSuccessData) {
          response.data = _this.config.transformSuccessData(response.data, response.header);
        }
        // @ts-ignore
        var okResponse = tslib_1.__assign(tslib_1.__assign({}, action), { type: success, response: response.data, effect: action.onSuccess });
        successInvoked = true;
        _this.collectResponse(okResponse);
        _this.next(okResponse);
        _this.triggerShowSuccess(okResponse, action.successText);
        return Promise.resolve(okResponse);
      })
      .catch(function (error) {
        if (successInvoked) {
          return Promise.reject(error);
        }
        var errorMessage;
        var httpStatus;
        var businessCode;
        if (error.statusCode) {              // **    statusCode --> status    **
          var transform = {
            httpStatus: error.statusCode,    // **    statusCode --> status    **
          };```

Here below is the Taro info:
👽 Taro v1.3.36

  Taro CLI 1.3.36 environment info:
    System:
      OS: macOS 10.14.6
      Shell: 5.3 - /bin/zsh
    Binaries:
      Node: 8.15.0 - /usr/local/bin/node
      Yarn: 1.19.1 - /usr/local/bin/yarn
      npm: 6.11.3 - /usr/local/bin/npm
    npmPackages:
      @tarojs/async-await: 1.3.36 => 1.3.36 
      @tarojs/components: 1.3.36 => 1.3.36 
      @tarojs/plugin-babel: 1.3.36 => 1.3.36 
      @tarojs/plugin-csso: 1.3.36 => 1.3.36 
      @tarojs/plugin-sass: 1.3.36 => 1.3.36 
      @tarojs/plugin-uglifyjs: 1.3.36 => 1.3.36 
      @tarojs/redux: 1.3.36 => 1.3.36 
      @tarojs/redux-h5: 1.3.36 => 1.3.36 
      @tarojs/router: 1.3.36 => 1.3.36 
      @tarojs/taro: 1.3.36 => 1.3.36 
      @tarojs/taro-alipay: 1.3.36 => 1.3.36 
      @tarojs/taro-h5: 1.3.36 => 1.3.36 
      @tarojs/taro-swan: 1.3.36 => 1.3.36 
      @tarojs/taro-tt: 1.3.36 => 1.3.36 
      @tarojs/taro-weapp: 1.3.36 => 1.3.36 
      @tarojs/webpack-runner: 1.3.36 => 1.3.36 
      eslint-config-taro: 1.3.36 => 1.3.36 
      eslint-plugin-taro: 1.3.36 => 1.3.36 
      nerv-devtools: ^1.5.6 => 1.5.6 
      nervjs: ^1.5.6 => 1.5.6 
      react: ^16.12.0 => 16.12.0 
      stylelint-config-taro-rn: 1.3.36 => 1.3.36 
      stylelint-taro-rn: 1.3.36 => 1.3.36 
      npmGlobalPackages:
      typescript: 3.7.2

**Comment by DevanXU** [comment]
BTW, the file we do the change locates in @redux-model/taro/libs/HttpService.js, line 75 & 77

**Comment by DevanXU** [comment]
Another question, is there any interception call for the response? e.g. We want to intercept response with business error code (http status == 200) for handling, but onRespondError only care for those responses with http status != 200. We tried "isSuccess" but seems like it didn't get called.

Thanks for your help.

Devan

**Comment by fwh1990** [comment]
> Another question, is there any interception call for the response? e.g. We want to intercept response with business error code (http status == 200) for handling, but onRespondError only care for those responses with http status != 200. We tried "isSuccess" but seems like it didn't get called.
> 
> Thanks for your help.
> 
> Devan

Yeah, using `isSuccess` is the right way to interrupt response. You can format response.data as you want, and just return false when you want to reject it.
```javascript
const service = new HttpService({
  ...
  isSuccess: (response) => {
      // Rejected now
      return false;
  },
  ...
});
```

**Comment by fwh1990** [comment]
> BTW, the file we do the change locates in @redux-model/taro/libs/HttpService.js, line 75 & 77

It seems that latest version of taro has changed the behavior. I'll check it out and fix it as soon as possible.

**Comment by fwh1990** [comment]
Fixed by commit [c3f83418fa50e0d565215b0fb824fdf7c9fc34de](https://github.com/redux-model/redux-model/commit/c3f83418fa50e0d565215b0fb824fdf7c9fc34de), upgrade package to latest version.

Thank for your asking, and feedback any suggestion or PR to me as you want.

**Comment by DevanXU** [comment]
> Fixed by commit [c3f83418fa50e0d565215b0fb824fdf7c9fc34de](https://github.com/redux-model/redux-model/commit/c3f83418fa50e0d565215b0fb824fdf7c9fc34de), upgrade package to latest version.
> 
> Thank for your asking, and feedback any suggestion or PR to me as you want.

We upgrade to the lastest verion 6.8.3 and it works. :)
Thanks.

**Comment by fwh1990** [comment]
Version 7.0 is released now, feel free to upgrade it.
